cborg/tests/Tests/FlatTerm.hs

```diff
@@ -0,0 +1,121 @@
+  ( testTree -- :: TestTree
+  ) where
+
+import           Codec.CBOR.FlatTerm
+
+import           Tests.Reference.Implementation
+import           Tests.Reference.Generators
+import qualified Data.ByteString      as BS
+import qualified Data.Text as T
+import           Numeric.Half
+
+import           Test.QuickCheck
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+--------------------------------------------------------------------------------
+
+
+newtype ValidFlat = ValidFlat [TermToken]
+  deriving (Eq, Show)
+
+instance Arbitrary ValidFlat where
+  arbitrary = ValidFlat . termToValidTermTokens <$> arbitrary
+
+newtype InValidFlat = InValidFlat [TermToken]
+  deriving (Eq, Show)
+
+instance Arbitrary InValidFlat where
+  arbitrary =
+    oneof [ genInValidFlatType1
+          , genInValidFlatType2
+          , genInValidFlatType3
+          , genInValidFlatType4
+          , InValidFlat . termToInValidTermTokens <$> arbitrary
+          ]
+
+-- InValid FlatTerms with Trailing elements
+genInValidFlatType1 :: Gen InValidFlat
+genInValidFlatType1 = do
+  ValidFlat ts <- arbitrary
+  return $ InValidFlat (TkBreak : ts)
+
+-- InValid FlatTerms with wrong ByteString order
+genInValidFlatType2 :: Gen InValidFlat
+genInValidFlatType2 = do
+  ValidFlat ts <- arbitrary
+  let ts' = TkBytesBegin : TkNull : ts
+  return $ InValidFlat ts'
+
+-- InValid FlatTerms with wrong String order
+genInValidFlatType3 :: Gen InValidFlat
+genInValidFlatType3 = do
+  ValidFlat ts <- arbitrary
+  let ts' = TkStringBegin : TkNull : ts
+  return $ InValidFlat ts'
+
+-- InValid FlatTerms with too long List / Map Length
+genInValidFlatType4 :: Gen InValidFlat
+genInValidFlatType4 = do
+  tok <- elements [TkListLen maxBound, TkMapLen maxBound]
+  let ts' = tok : repeat TkNull
+  return $ InValidFlat ts'
+
+
+-- Convert Term to List of Invalid TermTokens deliberately
+termToInValidTermTokens :: Term -> [TermToken]
+termToInValidTermTokens (TBytess ws)   = TkBytesBegin : ((TkBytes . BS.pack) <$> ws) ++ [TkBytesBegin]
+termToInValidTermTokens (TStrings ss)  = TkStringBegin : (TkString . T.pack <$> ss) ++ [TkBytesBegin]
+termToInValidTermTokens (TArray ts)    =
+  TkListLen (fromIntegral $ length ts) : (mconcat $ termToInValidTermTokens <$> ts) ++ [TkNull]
+termToInValidTermTokens (TArrayI ts)   =
+  TkListBegin : TkListBegin : (mconcat $ termToInValidTermTokens <$> ts) ++ [TkBreak]
+termToInValidTermTokens (TMap kvs)     =
+  let kvList = mconcat $ (\ (a, _) -> [a]) <$> kvs
+  in TkMapLen (fromIntegral $ length kvList) : (mconcat $ termToInValidTermTokens <$> kvList) ++ [TkNull]
+termToInValidTermTokens (TMapI kvs)    =
+  let kvList = mconcat $ (\ (a, _) -> [a]) <$> kvs
+  in TkMapBegin : TkStringBegin : (mconcat $ termToInValidTermTokens <$> kvList) ++ [TkBreak]
+termToInValidTermTokens x = TkStringBegin : TkNull : TkBreak : termToValidTermTokens x
+
+
+-- Convert Term to List of Valid TermTokens
+termToValidTermTokens :: Term -> [TermToken]
```

Comment:
This could be replaced by `toFlatTerm . encodeTerm` couldn't it?

Reply:
I don't think it could be done so easily because of type mismatch, as the `Term` type here is referring to `Tests.Reference.Implementaion` (for which we have defined arbitrary instance) and the `Term` type  expected by `encodeTerm` is from `Codec.CBOR.Term`. Also the `encodeTerm` defined in `Tests.Reference.Implementaion` has type `Encoder Term` i.e. `Term -> [Word8]`

Would you like to suggest something?

Reply:
Done!